### PR TITLE
Read Response with timestamped SenML

### PR DESCRIPTION
**Pull request by mgdlkundera**
Implementation of read response with timestamped SenML

This aims to implement : https://github.com/eclipse-leshan/leshan/issues/1553

### CODE REVIEW

File: leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/lockstep/LockStepTest.java [hunk 1]
```diff
@@ -315,7 +321,7 @@ public ReadResponse call() throws Exception {
 
         // Acknowledge the response
         client.expectRequest().storeMID("R").go();
-        client.sendEmpty(Type.ACK).loadMID("R").go();
+        client.sendEmpty(Type.ACK, ContentFormat.TEXT).loadMID("R").go();
```
Comment (sbernard31) [inline leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/lockstep/LockStepTest.java:324]: I'm not sure I get why adding content format when sending Empty ACK ?

File: leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/lockstep/LockStepTest.java [hunk 2]
```diff
@@ -370,7 +376,7 @@ public void async_send_with_acknowleged_request_without_response(String givenSer
...
```
Comment (sbernard31) [inline leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/lockstep/LockStepTest.java:379]: same here

File: leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/lockstep/LockStepTest.java [hunk 3]
```diff
@@ -428,4 +434,47 @@ public void register_deregister_observe(String givenServerEndpointProvider) thro
...
+        // --------------------------------------------------------------------------------------------------------------
+
+        // Send read REQUEST
+        Future<ReadResponse> future = Executors.newSingleThreadExecutor().submit(new Callable<ReadResponse>() {
+            @Override
+            public ReadResponse call() throws Exception {
+                // send a request with 3 seconds timeout
+                return server.send(registration, new ReadRequest(ContentFormat.SENML_JSON, 1), 2000);
+            }
+        });
```
Comment (sbernard31) [inline leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/lockstep/LockStepTest.java:469]: here maybe you can use java lambda notation ?

File: leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/lockstep/LockStepTest.java [hunk 4]
```diff
@@ -428,4 +434,47 @@ public void register_deregister_observe(String givenServerEndpointProvider) thro
...
+        byte[] payload = encoder.encodeTimestampedNodes(timestampedNodes, ContentFormat.SENML_JSON,
+                client.getLwM2mModel());
+        // --------------------------------------------------------------------------------------------------------------
+
+        // Send read REQUEST
+        Future<ReadResponse> future = Executors.newSingleThreadExecutor().submit(new Callable<ReadResponse>() {
+            @Override
+            public ReadResponse call() throws Exception {
+                // send a request with 3 seconds timeout
+                return server.send(registration, new ReadRequest(ContentFormat.SENML_JSON, 1), 2000);
```
Comment (sbernard31) [inline leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/lockstep/LockStepTest.java:467]: 2 seconds timeout is maybe too high ? 1 is enough

File: leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/lockstep/LockStepTest.java [hunk 5]
```diff
@@ -428,4 +434,47 @@ public void register_deregister_observe(String givenServerEndpointProvider) thro
...
+            }
+        });
+
+        client.expectRequest().storeToken("TKN").storeMID("MID").go();
+
+        client.sendResponse(Type.ACK, ResponseCode.CONTENT, ContentFormat.SENML_JSON).loadMID("MID").loadToken("TKN")
+                .payload(payload).go();
+
+        ReadResponse response = future.get(2000, TimeUnit.MILLISECONDS);
```
Comment (sbernard31) [inline leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/lockstep/LockStepTest.java:476]: 2 seconds timeout is maybe too high ? 1 is enough

(why not just using 1 and TimeUnit.SECONDS)

File: leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/endpoint/ServerCoapMessageTranslator.java [hunk 6]
```diff
@@ -67,10 +70,39 @@ public Request createCoapRequest(ClientProfile clientProfile, DownlinkRequest<?
...
+                        timestampedNode, null, coapResponse);
+                return (T) readResponse;
+
+            } else {
+                return builder.getResponse();
+            }
+        } else {
+            return builder.getResponse();
+        }
+
```
Comment (sbernard31) [inline leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/endpoint/ServerCoapMessageTranslator.java:105]: Are you sure this code is still needed ? this sounds to be done in `LwM2mResponseBuilder`, or I missed something  ?

File: leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/request/LwM2mResponseBuilder.java [hunk 7]
```diff
@@ -117,12 +118,30 @@ public LwM2mResponseBuilder(Request coapRequest, Response coapResponse, String c
...
+            if (request.getContentFormat() != null)
+                contentFormat = request.getContentFormat();
+            TimestampedLwM2mNode timestampedNode = null;
+            try {
+                List<TimestampedLwM2mNode> timestampedNodes = decoder.decodeTimestampedData(coapResponse.getPayload(),
+                        contentFormat, request.getPath(), model);
+                timestampedNode = timestampedNodes != null ? timestampedNodes.get(0) : null;
+            } catch (CodecException e) {
+                LOG.debug(String.format("Unable to decode response payload of request [%s]", request));
+            }
```
Comment (sbernard31) [inline leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/request/LwM2mResponseBuilder.java:135]: Maybe better to create a dedicated method to decode payload.

File: leshan-tl-javacoap-server/src/main/java/org/eclipse/leshan/transport/javacoap/server/request/LwM2mResponseBuilder.java [hunk 8]
```diff
@@ -117,14 +118,38 @@ public LwM2mResponseBuilder(CoapResponse coapResponse, CoapRequest coapRequest,
...
             // handle error response:
-            lwM2mresponse = new ReadResponse(toLwM2mResponseCode(coapResponse.getCode()), null,
-                    coapResponse.getPayloadString(), coapResponse);
+            if (timestampedNode != null && timestampedNode.isTimestamped()) {
+                lwM2mresponse = new ReadResponse(toLwM2mResponseCode(coapResponse.getCode()), null, timestampedNode,
+                        coapResponse.getPayloadString(), coapResponse);
+            } else {
+                lwM2mresponse = new ReadResponse(toLwM2mResponseCode(coapResponse.getCode()), null, null,
+                        coapResponse.getPayloadString(), coapResponse);
+            }
```
Comment (sbernard31) [inline leshan-tl-javacoap-server/src/main/java/org/eclipse/leshan/transport/javacoap/server/request/LwM2mResponseBuilder.java:142]: the comment seems to says this is about error response why trying to decode timestamped node here ? :thinking:

File: leshan-tl-javacoap-server/src/main/java/org/eclipse/leshan/transport/javacoap/server/request/LwM2mResponseBuilder.java [hunk 9]
```diff
@@ -117,14 +118,38 @@ public LwM2mResponseBuilder(CoapResponse coapResponse, CoapRequest coapRequest,
...
+
+        TimestampedLwM2mNode timestampedNode = null;
+        try {
+            List<TimestampedLwM2mNode> timestampedNodes = decoder.decodeTimestampedData(
+                    coapResponse.getPayload().getBytes(), contentFormat, request.getPath(), model);
+            timestampedNode = timestampedNodes != null ? timestampedNodes.get(0) : null;
+        } catch (CodecException e) {
+            LOG.debug(String.format("Unable to decode response payload of request [%s]", request));
+        }
+
```
Comment (sbernard31) [inline leshan-tl-javacoap-server/src/main/java/org/eclipse/leshan/transport/javacoap/server/request/LwM2mResponseBuilder.java:133]: Probably better to create dedicated method here too and call it only in : 

// handle success response

section only

**Review by sbernard31 — COMMENTED** [review]
My first quick review.

I didn't checkout / test the code yet.


Do you want to handle timestamped value for `ObserveResponse` too ? 
(I mean for observe response, not notification (notification as already timestamp value)

File: leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/lockstep/LockStepTest.java (hunk 4 above)
Comment (mgdlkundera) [reply leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/lockstep/LockStepTest.java:467]: We should leave a 2 seconds timeout, with 1 s we can't get a response

Comment (sbernard31) [reply leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/lockstep/LockStepTest.java]: 1s should be largely enough to get a response using localhost in tests.

So I read the test again and I see that the comment says : 
 - 3 seconds
 - and the timeout value parameter value is `1`
 
 But the javadoc says : 
```
 * @param timeoutInMs The global timeout to wait in milliseconds
```

So it should be 3 seconds and 3000 or 1 seconds and 1000

**Comment by sbernard31** [comment]
I rebase the code on `master` and push some fixes.

You could review it if wanted. (I advice to review commit by commit)

We still need to decide : 
> Do you want to handle timestamped value for ObserveResponse too ?
(I mean for observe response, not notification (notification as already timestamp value)

I didn't have time to test about reducing timeout value but I pretty sure that 3 seconds is not needed

**Comment by mgdlkundera** [comment]
I reviewed the code and everything looks fine. I also pushed commit with timestamped for Observe Response.

**Comment by sbernard31** [comment]
> I reviewed the code and everything looks fine. I also pushed commit with timestamped for Observe Response.

:pray: 

I started to work on it again. I should be able to push some code tomorrow.

**Comment by sbernard31** [comment]
@mgdlkundera, I added some more fix/features.

You could review/test it if wanted. (I advice to review commit by commit)

@mjviljan is possible to test it ? (let me know if this is a problem for you to test a PR)

**Comment by mjviljan** [comment]
> @mjviljan is possible to test it ? (let me know if this is a problem for you to test a PR)

Yes, I can test this also. Unfortunately only some time next week, so if you want to merge this before that, don't wait for me.

But otherwise I'll come back to this once I've tried it out.

**Comment by sbernard31** [comment]
@mjviljan thx for letting me know that. We can wait for next week.

**Comment by mjviljan** [comment]
Hi, sorry for the delay. I tested this now for [my case](https://github.com/eclipse-leshan/leshan/issues/1616). Unfortunately this doesn't fix that — but I guess that's just (more) proof that the data I receive from the device may not be really LwM2M standard-compliant.

Anyway, thanks for suggesting me to test this! At least I now have some new knowledge to continue with.

**Comment by mgdlkundera** [comment]
I tested it and I also look through the code and it is ok.

**Comment by sbernard31** [comment]
> I tested it and I also look through the code and it is ok.

Thx :pray: 

I rebased it on `master`, then squashed all commit in one (dceda02a1fd6f99c93b6742c49fc98e5cdb89ddc), then integrated in `master`.
It will be available in 2.0.0-M15.

**Pull request closed without merging**